### DISCUSSION TITLE
Cross compile arm64 on x86 arch

### DISCUSSION THREAD

**Discussion by vincentdchan** [ocaml/dune#5374] [Q&A]
## Desired Behavior

Apple has released M1 sillicon for a while. It's inconvient to compile OCaml code for different platforms.
Such as compiling M1 version on a x86 mac, or compiling x86 on a M1 mac. Is that possible for OCaml?

Both Go and Rust support this feature.

**Comment by rgrinberg** [comment]
It's possible. It does require you to install a host and target compiler though. See documentation: https://dune.readthedocs.io/en/stable/cross-compilation.html

**Reply by vincentdchan** [reply]
I don't know how to choose the `arch` of target from the document you presented.

**Comment by lord** [comment]
One gross and probably overcomplicated (but fully working!) solution I've found is that, since ARM macs can just run x86_64 binaries, you don't need to cross compile at all. Instead, an alternate intel installation of ocaml can build intel binaries. Very roughly:

```
export OPAMROOT=$HOME/Desktop/opam-x86_64/root
export DUNE_BUILD_DIR=_build_intel
arch -x86_64 /bin/bash -c "sh <(curl -fsSL https://opam.ocaml.org/install.sh)"  # install to ~/Desktop/opam-x86_64 when prompted
mkdir ~/Desktop/opam-x86_64/root
arch -x86_64 ~/Desktop/opam-x86_64/opam init --no-setup
arch -x86_64 ~/Desktop/opam-x86_64/opam install dune
arch -x86_64 ~/Desktop/opam-x86_64/opam exec dune -- build <list targets here>
```
